README.rst
==================================
Bayesian estimation of proportions
==================================

.. image:: https://img.shields.io/pypi/v/bayespropestimation.svg
        :target: https://pypi.python.org/pypi/bayespropestimation


Simple class and methods for the Bayesian estimation and comparison of proportions.

* Free software: MIT license

Features
--------

* Estimates the posterior distribution of the mean parameter for two binomial samples, A and B.
* Estimates of the posterior distribution of difference in mean parameters for two binomial samples, A and B.
* Provides summary statistics and visualisations for the estimated parameters.
* The prior distribution, sample count, random seed, credible intervals and parameter names can all be customised.


============
Installation
============

Stable release
--------------

To install Bayesian estimation of proportions, run this command in your terminal:

.. code-block:: console

    $ pip install bayespropestimation

This is the preferred method to install the module, as it will always install the most recent stable release.

If you don't have `pip`_ installed, this `Python installation guide`_ can guide
you through the process.

.. _pip: https://pip.pypa.io
.. _Python installation guide: http://docs.python-guide.org/en/latest/starting/installation/

From sources
------------

You can either clone the public repository:

.. code-block:: console

    $ git clone git://github.com/oli-chipperfield/bayespropestimation

Or download the `tarball`_:

.. code-block:: console

    $ curl -OJL https://github.com/oli-chipperfield/bayespropestimation/tarball/master

Once you have a copy of the source, you can install it with:

.. code-block:: console

    $ python setup.py install

.. _Github repo: https://github.com/oli-chipperfield/bayespropestimation
.. _tarball: https://github.com/oli-chipperfield/bayespropestimation/tarball/master

===========
Methodology
===========

See `notebook <https://github.com/oli-chipperfield/bayespropestimation/blob/master/docs/bayespropestimation_basis.ipynb>`_ for details.

=====
Usage
=====

To use Bayesian estimation of proportions in a project

.. code-block:: python

    import bayespropestimation

Simple example
--------------

To carry out a simple estimation of the posterior density of two samples and an estimate of the difference, using an uninformative prior.  Import the `BayesProportionsEstimation` class.

.. code-block:: python

    from bayespropestimation.bayespropestimation import BayesProportionsEstimation

Define data from samples A and B as two lists of format `[successes, trials]` and initialise the `BayesProportionsEstimation` class.

.. code-block:: python

    a = [10, 50]
    b = [20, 50]
    ExampleBayes = BayesProportionsEstimation(a, b)

Posterior densities are estimated when the class is intialised. There are five methods for accessing information about the draws from simulations of the posterior densities.

.. code-block:: python

    ExampleBayes.get_posteriors()
    # Returns tuple of samples from the posterior distributions for parameters

.. code-block:: python

    ExampleBayes.hdi_summary()
    # Returns dataframe of the high-density-interval (HDI), maximum-a-posteriori (MAP) and mean of samples from the posteriores

.. image:: https://github.com/oli-chipperfield/bayespropestimation/blob/master/images/example_hdi.png

.. code-block:: python

    ExampleBayes.quantile_summary()
    # Returns dataframe of quantiles and mean of the posterior densities of samples for parameters

.. image:: https://github.com/oli-chipperfield/bayespropestimation/blob/master/images/example_quantile.png

.. code-block:: python

    ExampleBayes.infer_delta_probability()
    # Returns probability estimate of the delta parameter being greater than 0, plus an aid to inference.  
    #Includes an optional print out of the probability and inference.
    
    'The probability that theta_b is greater than theta_a is 98.63%. Therefore theta_b is almost certainly greater than theta_a.'
    '(0.9863, 'almost certainly')'

.. code-block:: python

    ExampleBayes.infer_delta_bayes_factor()
    # Returns the Bayes factor of the hypothesis that P(theta_b > theta_a | D) where D is the data, plus an aid to inference.  
    # Includes an optional print out of the Bayes factor and inference.

    'The calculated bayes factor for the hypothesis that theta_b is greater than theta_a versus the hypothesis that theta_a is greater than theta_a is 71.993. Therefore the strength of evidence for this hypothesis is very strong.'
    '(71.99270072992677, 'very strong')'

.. code-block:: python

    ExampleBayes.posterior_plot
    # Returns KDE plots of samples from the posterior densities of the parameters

.. image:: https://github.com/oli-chipperfield/bayespropestimation/blob/master/images/example_posterior_plot.png

To see how to use non-default parameters, refer to the `usage guid <https://github.com/oli-chipperfield/bayespropestimation/blob/master/docs/bayespropestimation_usage.ipynb>`_ or refer to the doc-strings in the `source <https://github.com/oli-chipperfield/bayespropestimation/bayespropestimation/bayespropestimation.py>`_.

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage

.. highlight:: shell


HISTORY.rst
=======
History
=======

0.1.0 (2020-06-20)
------------------

* First release on PyPI.

0.1.7 (Sep 2020)
----------------

* Migrate to plotly visualisations
* Add MAP and HDI methods
* Add additional helper functions

0.9.0 (2020-10-10)
------------------

* Almost 'final' version
* Improve readmes

0.9.1 (2020-11-08)
------------------

* Almost 'final' version
* Update requirements
* Run black

tests/test_bayespropestimation.py
#!/usr/bin/env python
import pytest
import numpy as np
import pandas as pd
from bayespropestimation.bayespropestimation import BayesProportionsEstimation
from bayespropestimation.bayesprophelpers import _calculate_kde
from bayespropestimation.bayesprophelpers import _calculate_map
from bayespropestimation.bayespropplotters import _get_centre_lines
from bayespropestimation.bayespropplotters import _get_intervals 
from bayespropestimation.bayespropplotters import _make_density_go 
from bayespropestimation.bayespropplotters import _make_histogram_go 
from bayespropestimation.bayespropplotters import _make_area_go
from bayespropestimation.bayespropplotters import _make_line_go
from bayespropestimation.bayespropplotters import _make_delta_line


def compare_dictionaries(p, z):
    k = list(z.keys())
    r = []
    for i in k:
        r.append(np.any(p[i], z[i]))
    return np.any(r)

# Define Initialisation fixtures

@pytest.fixture
def make_a_list():
    return [10, 50]

@pytest.fixture
def make_a_numpy():
    return np.array([10, 50])

@pytest.fixture
def make_a_series():
    return pd.DataFrame({'a': [10, 50]})['a']

@pytest.fixture
def make_b_list():
    return [20, 50]

@pytest.fixture
def make_b_numpy():
    return np.array([20, 50])

@pytest.fixture
def make_b_series():
    return pd.DataFrame({'a': [20, 50]})['a']

@pytest.fixture
def make_explicit_prior():
    return 0.1

@pytest.fixture
def make_explicit_n():
    return 3

@pytest.fixture
def make_explicit_seed():
    return 1000

# Define bad inputs

@pytest.fixture
def make_a_str():
    return 'foo'

@pytest.fixture
def make_a_bad_proportion():
    return [50, 10]

@pytest.fixture
def make_invalid_prior():
    return -1

@pytest.fixture
def make_invalid_n():
    return -1

@pytest.fixture
def make_invalid_seed():
    return 1.2

# Define fixtures for results testing

@pytest.fixture
def make_quantile_summary_results():
    return np.array([[0.10594123533474246, 0.20127340652543865, 0.32294649108461715],
                     [0.2744412955144984, 0.3999900504080393, 0.5396945573617622],
                     [0.019031409978063265, 0.1974645057248544, 0.3708045045911511]])

@pytest.fixture
def make_get_posterior_results():
    return np.array([[0.15608103, 0.17790474, 0.24626163],
                    [0.35279345, 0.3636633 , 0.51329111],
                    [0.19671241, 0.18575856, 0.26702948]])

@pytest.fixture
def make_hdi_summary_results():
    return np.array([[0.1038032606788284, 0.19275171884176634, 0.320526260831967],
                     [0.2691767369659035, 0.3986262017019868, 0.5338200338157294],
                     [0.0242652248218867, 0.19609368609983302, 0.3745174565063844]])

# Define fixtures for bayeprophelpers testing

@pytest.fixture
def make_draw():
    np.random.seed(1000)
    return np.random.beta(2, 5, 100)

@pytest.fixture
def _calculate_kde_results():
    return (np.array([0.03857532, 0.43082818, 0.82308105]),
            np.array([1.14415602, 1.12407264, 0.11679198]))

@pytest.fixture
def _calculate_map_results():
    return 0.038575315797957296

# Run helper tests

def test__calculate_kde_returns_correct_values(make_draw, _calculate_kde_results):
    x, kde_density = _calculate_kde(make_draw, num=3)
    assert np.allclose(x, _calculate_kde_results[0])
    assert np.allclose(kde_density, _calculate_kde_results[1])

def test__calculate_map_results(make_draw, _calculate_map_results):
    assert np.isclose(_calculate_map_results, _calculate_map(make_draw, num = 3))

# Run initialisation tests

def test_BayesProportionsEstimation_with_make_a_list_and_make_b_list_initialises(make_a_list, make_b_list):
    try:
        BayesProportionsEstimation(a=make_a_list, b=make_b_list)
    except:
        raise pytest.fail()

def test_BayesProportionsEstimation_with_make_a_numpy_and_make_b_numpy_initialises(make_a_numpy, make_b_numpy):
    try:
        BayesProportionsEstimation(a=make_a_numpy, b=make_b_numpy)
    except:
        raise pytest.fail()

def test_BayesProportionsEstimation_with_make_a_series_and_make_b_series_initialises(make_a_series, make_b_series):
    try:
        BayesProportionsEstimation(a=make_a_series, b=make_b_series)
    except:
        raise pytest.fail()

def test_BayesProportionsEstimation_with_make_explicit_prior_alpha_and_make_explicit_prior_beta_intialises(make_a_list, make_b_list, make_explicit_prior):
    try:
        BayesProportionsEstimation(a=make_a_list, b=make_b_list, prior_alpha=make_explicit_prior)
    except:
        raise pytest.fail()

def test_BayesProportionsEstimation_with_mmake_explicit_n_initialises(make_a_list, make_b_list, make_explicit_n):
    try:
        BayesProportionsEstimation(a=make_a_list, b=make_b_list, n=make_explicit_n)
    except:
        raise pytest.fail()

def test_BayesProportionsEstimation_with_mmake_explicit_seed_initialises(make_a_list, make_b_list, make_explicit_seed):
    try:
        BayesProportionsEstimation(a=make_a_list, b=make_b_list, seed=make_explicit_seed)
    except:
        raise pytest.fail()

# Run ValueError tests

def test_BayesProportionsEstimation_with_a_str_returns_ValueError(make_a_str, make_b_list):
    with pytest.raises(ValueError) as e:
        BayesProportionsEstimation(a=make_a_str, b=make_b_list)
    assert str(e.value) == "type(a).__name__ and/or type(b).__name__ must be 'list', 'ndarray' or 'DataFrame'"

def test_BayesProportionsEstimation_with_a_bad_proportion_returns_ValueError(make_a_bad_proportion, make_b_list):
    with pytest.raises(ValueError) as e:
        BayesProportionsEstimation(a=make_a_bad_proportion, b=make_b_list)
    assert str(e.value) == "the count of successes for a and/or b exceeds the number of trials"

def test_BayesProportionsEstimation_with_make_invalid_prior_returns_ValueError(make_a_list, make_b_list, make_invalid_prior):
    with pytest.raises(ValueError) as e:
        BayesProportionsEstimation(a=make_a_list, b=make_b_list, prior_alpha=make_invalid_prior)
    assert str(e.value) == "the prior_alpha and/or prior_beta parameters must be > 0"

def test_BayesProportionsEstimation_with_make_invalid_n_returns_ValueError(make_a_list, make_b_list, make_invalid_n):
    with pytest.raises(ValueError) as e:
        BayesProportionsEstimation(a=make_a_list, b=make_b_list, n=make_invalid_n)
    assert str(e.value) == "n must be a positive integer"

def test_BayesProportionsEstimation_with_make_invalid_seed_returns_ValueError(make_a_list, make_b_list, make_invalid_seed):
    with pytest.raises(ValueError) as e:
        BayesProportionsEstimation(a=make_a_list, b=make_b_list, seed=make_invalid_seed)
    assert str(e.value) == "seed must be a positive integer or None"

# Run results tests

def test_BayesProportionsEstimation_quantile_summary_returns_correct_results(make_a_list, make_b_list, make_explicit_seed, make_quantile_summary_results):
    test = np.array(BayesProportionsEstimation(a=make_a_list, b=make_b_list, seed=make_explicit_seed).quantile_summary())[:,0:3]
    assert np.array_equal(test, make_quantile_summary_results)

def test_BayesProportionsEstimation_hdi_summary_returns_correct_results(make_a_list, make_b_list, make_explicit_seed, make_quantile_summary_results, make_hdi_summary_results):
    test = np.array(BayesProportionsEstimation(a=make_a_list, b=make_b_list, seed=make_explicit_seed).hdi_summary())[:,0:3]
    assert np.array_equal(test, make_hdi_summary_results)

def test_BayesProportionsEstimation_get_posteriors_returns_correct_results(make_a_list, make_b_list, make_explicit_n, make_explicit_seed, make_get_posterior_results):
    test = BayesProportionsEstimation(make_a_list, make_b_list, n=make_explicit_n, seed=make_explicit_seed).get_posteriors()
    test = np.array(pd.DataFrame(test))
    assert np.allclose(test, make_get_posterior_results)

# RUn bayespropplotters tests

def test__get_centre_lines_runs_without_error(make_draw):
    try:
        _get_centre_lines(make_draw, method='hdi')
    except:
        raise pytest.fail()

def test__get_intervals_runs_with_hdi_without_error(make_draw):
    try:
        _get_intervals(make_draw, method='hdi', bounds=0.95)
    except:
        raise pytest.fail()

def test__get_intervals_runs_with_quantile_without_error(make_draw):
    try:
        _get_intervals(make_draw, method='quantile', bounds=[0.025, 0.975])
    except:
        raise pytest.fail()

def test__make_density_go_without_error(make_draw):
    try:
        _make_density_go(make_draw, name='dummy')
    except:
        raise pytest.fail()

def test__make_histogram_go_without_error(make_draw):
    try:
        _make_histogram_go(make_draw, name='dummy')
    except:
        raise pytest.fail()

def test__make_line_go_with_without_error(make_draw):
    try:
        cl = _get_centre_lines(make_draw, method='hdi')
        _make_line_go(cl, name='dummy')
    except:
        raise pytest.fail()

def test__make_area_go_without_error(make_draw):
    try:
        intervals = _get_intervals(make_draw, method='hdi', bounds=0.95)
        _make_area_go(intervals, name='dummy')
    except:
        raise pytest.fail()

bayespropestimation/bayespropplotters.py
import numpy as np
import pandas as pd
import scipy as scipy
import arviz as az
import plotly.graph_objects as go

from bayespropestimation.bayesprophelpers import _calculate_kde, _calculate_map

def _get_centre_lines(draws, method):
    # Derives map or median or mean for plotting purposes   
    if method == 'hdi':
        cl = _calculate_map(draws)
    elif method == 'quantile':
        cl = np.quantile(draws, 0.5)
    x, kde_density = _calculate_kde(draws, num = 100)
    best_y = kde_density[np.argmin(np.abs(x - cl))]
    return {'x': cl, 'y': best_y}

def _get_intervals(draws, method, bounds):
    # Derives HDI or credible intervals for plotting purposes
    if method == 'hdi':
        il = az.hdi(draws, bounds)
    elif method == 'quantile':
        il = np.quantile(draws, bounds)
    x, kde_density = _calculate_kde(draws, num = 100)
    subx = x[(x > il[0]) & (x < il[1])]
    kde_density = kde_density[(x > il[0]) & (x < il[1])]
    return {'x': subx, 'y': kde_density}

def _make_density_go(draws, name, col='#000000'):
    # Makes a KDE density graph object
    x, kde_density = _calculate_kde(draws, num = 100)
    graphobj = go.Scatter(x=x,
                          y=kde_density,
                          line={'color': col},
                          name=name)
    return graphobj

def _make_histogram_go(draws, name, col='#000000'):
    # Makes a histogram graph object
    graphobj = go.Histogram(x=draws, 
                            opacity=0.2, 
                            histnorm='probability density',
                            marker_color=col,
                            name=name)
    return graphobj

def _make_area_go(kde_object, name, col='#000000'):
    # Makes area under curve graph object
    graphobj = go.Scatter(x=kde_object['x'],
                          y=kde_object['y'],
                          fill='tozeroy',
                          line={'color': col},
                          opacity=0.5,
                          name=name)
    return graphobj

def _make_line_go(line_object, name, col='#000000'):
    # Make line graph object
    graphobj = go.Scatter(x=[line_object['x'], line_object['x']],
                          y=[0, line_object['y']],
                          mode='lines',
                          line={'color': col, 'dash': 'dash'},
                          name=name)
    return graphobj

def _make_delta_line(draws, delta_line, col='#d62728'):
    # Make line dictionary object for the delta posterior
    x, kde_density = _calculate_kde(draws, num = 100)
    maxy = np.max(kde_density) * 1.2    
    linedict = {'type': 'line',
                'x0': delta_line,
                'x1': delta_line,
                'y0': 0,
                'y1': maxy,
                'xref': 'x3',
                'yref': 'y3',
                'line': dict(color=col, dash='dot')}
    return linedict










bayespropestimation/bayespropestimation.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy as scipy
import arviz as az

from plotly.subplots import make_subplots
from bayespropestimation.bayesprophelpers import _calculate_kde
from bayespropestimation.bayesprophelpers import _calculate_map
from bayespropestimation.bayespropplotters import _get_centre_lines
from bayespropestimation.bayespropplotters import _get_intervals 
from bayespropestimation.bayespropplotters import _make_density_go 
from bayespropestimation.bayespropplotters import _make_histogram_go 
from bayespropestimation.bayespropplotters import _make_area_go
from bayespropestimation.bayespropplotters import _make_line_go
from bayespropestimation.bayespropplotters import _make_delta_line

class BayesProportionsEstimation:

    def __init__(self, a, b, prior_alpha=0.5, prior_beta=0.5, n=10000, seed=None):
        '''
        Initialises the BayesProportionsEstimation class and samples from the posterior distribution
        Parameters
        ----------
        a: list, ndarray or Series [successes, trials]:  array describing results from sample a
        b: list, ndarray or Series [successes, trials]:  array describing results from sample b
        prior_alpha: float, alpha parameter for the Beta prior distribution, default = 0.5 (Jeffreys prior)
        prior_beta: float, beta parameter for the Beta prior distribution, default = 0.5 (Jeffreys prior)
        m: integer, number of samples to take from the posterior distribution, default = 10000
        seed: integer, set random seed at the start of the initialisation, default = None
        '''   
        self.a = a
        self.b = b
        self.prior_alpha = prior_alpha
        self.prior_beta = prior_beta
        self.n = n  
        self.seed = seed
        self._check_inputs()
        self._sample_posteriors()

    def _check_inputs(self):
        # Checks that parameters are in the correct format
        types = ['list', 'ndarray', 'Series']
        if ((type(self.a).__name__ not in types) or (type(self.b).__name__ not in types)):
            raise ValueError("type(a).__name__ and/or type(b).__name__ must be 'list', 'ndarray' or 'DataFrame'")
        if ((self.a[0] > self.a[1]) or (self.b[0] > self.b[1])):
            raise ValueError("the count of successes for a and/or b exceeds the number of trials")
        if ((self.prior_alpha <= 0) or (self.prior_beta <= 0)):
            raise ValueError("the prior_alpha and/or prior_beta parameters must be > 0")
        if self.n <= 0:
            raise ValueError("n must be a positive integer")
        if self.seed is not None and str(self.seed).isdigit() == False:
            raise ValueError("seed must be a positive integer or None")

    def _posterior_function(self, d):
        # Defines the posterior
        return np.random.beta(d[0] + self.prior_alpha, d[1] - d[0] + self.prior_beta, self.n)

    def _sample_posteriors(self):
        # Draws from posterior
        np.random.seed(self.seed)
        a_draw = self._posterior_function(self.a)
        b_draw = self._posterior_function(self.b)
        d_draw = b_draw - a_draw
        self.a_draw = a_draw
        self.b_draw = b_draw
        self.d_draw = d_draw

    def get_posteriors(self):
        '''
        Retrieves random draws from the posterior
        Returns
        -------
        tuple:
            - np.array[n] draws from the posterior of theta_a
            - np.array[n] draws from the posterior of theta_b
            - np.array[n] draws from the posterior of theta_b minus draws from the posterior of theta_a
        '''
        return (self.a_draw, 
                self.b_draw, 
                self.d_draw)

    def _calculate_quantiles(self, d, mean, quantiles):
        # Calculate mean and quantiles
        q = np.quantile(d, quantiles)
        if mean is True:
            q = np.append(q, np.mean(d))    
        return q

    def quantile_summary(self, mean=True, quantiles=[0.025, 0.5, 0.975], names=None):
        '''
        Summarises the properties of the estimated posterior using quantiles
        Parameters
        ----------
        mean:  boolean, default True, calculates the mean of the draws from the posterior.  Default True
        quantiles: list, calculates the quantiles of the draws from the posterior.  Default [0.025, 0.5, 0.975]
        names:  list of length 3, parameter names in order: a, b, b-a.  Default ['theta_a', 'theta_b', 'delta']
        Returns
        -------
        pd.DataFrame:  
            'theta_a':  summaries of the posterior of theta_a
            'theta_b':  summaries of the posterior of theta_b
            'delta':  summaries of the posterior of theta_b - theta_a
        '''
        if quantiles is None:
            raise ValueError("quantiles must be a list of length > 0")      
        draws = [self.a_draw, self.b_draw, self.d_draw]
        if names is None:
            names = [
                    'theta_a',
                    'theta_b',
                    'delta'
                    ]
        if len(names) > 3:
            raise ValueError('names must be a list of length 3')
        q = []
        for i in draws:
            q.append(self._calculate_quantiles(i, mean, quantiles))
        df = pd.DataFrame(np.array(q))
        if mean is True:
            df.columns = list(map(str, quantiles)) + ['mean']
        else:
            df.columns = list(map(str, quantiles)) 
        df['parameter'] = names
        return df

    def _calculate_hdi_and_map(self, d, mean, interval):
        # Calculate HDI interval and MAP
        q = az.hdi(d, hdi_prob=interval)
        m = _calculate_map(d)
        q = np.array([q[0], m, q[1]])
        if mean is True:
            q = np.append(q, np.mean(d))
        return q

    def hdi_summary(self, mean=True, interval=0.95, names=None):
        '''
        Summarises the properties of the estimated posterior using the MAP and HDI
        Parameters
        ----------
        mean:  boolean, calculates the mean of the draws from the posterior.  Default True
        interval: float, defines the HDI interval.  Default = 0.95 (i.e. 95% HDI interval)
        names:  list of length 3, parameter names in order: a, b, b-a.  Default ['theta_a', 'theta_b', 'delta']
        Returns
        -------
        pd.DataFrame:  
            'theta_a':  summaries of the posterior of theta_a
            'theta_b':  summaries of the posterior of theta_b
            'delta':  summaries of the posterior of theta_b - theta_a
        '''
        if interval is None or interval <= 0 or interval >= 1:
            raise ValueError("interval must be a float > 0 and < 1")      
        draws = [self.a_draw, self.b_draw, self.d_draw]
        if names is None:
            names = [
                    'theta_a',
                    'theta_b',
                    'delta'
                    ]
        if len(names) > 3:
            raise ValueError('names must be a list of length 3')
        q = []
        for i in draws:
            q.append(self._calculate_hdi_and_map(i, mean, interval))        
        df = pd.DataFrame(np.array(q))
        col_names = ['%.5g' % ((1 - interval) / 2), 'MAP', '%.5g' % (interval + ((1 - interval) / 2))]
        if mean is True:
            df.columns = col_names + ['mean']
        else:
            df.columns = col_names
        df['parameter'] = names
        return df    

    def _probability_interpretation_guide(self, p):
        # Interpretation guide for probabilities using: 
        # https://www.cia.gov/library/center-for-the-study-of-intelligence/csi-publications/books-and-monographs/sherman-kent-and-the-board-of-national-estimates-collected-essays/6words.html
        if p >= 0 and p <= 0.13:
            i = 'almost certainly not'
        elif p > 0.13 and p <= 0.4:
            i = 'probably not'
        elif p > 0.4 and p <= 0.6:
            i = 'about equally likely'
        elif p > 0.6 and p <= 0.86:
            i = 'probably'
        elif p > 0.86 and p <= 1:
            i = 'almost certainly'
        else:
            raise ValueError('p must be >= 0 and <= 1')
        return i

    def _print_inference_probability(self, p, i, direction, value, names):
        # Combines inference values into a readable string
        s = 'The probability that ' + names[1] + ' is ' + direction + ' ' + names[0]
        if value != 0:
            s = s + ' by more than ' + str(value)
        s = s + ' is ' + ('%.5g' % (p * 100)) + '%.'
        s = s + ' Therefore ' + names[1] + ' is ' + i + ' ' + direction + ' ' + names[0]
        if value != 0:
            s = s + ' by more than ' + str(value)
        s = s + '.'
        return s

    def infer_delta_probability(self, direction='greater than', value=0, print_inference=True, names = None):
        '''
        Provides a guide to making inferences on the posterior delta, based on proportion of
        draws to the right or left of a given value. 
        Parameters
        ----------
        direction: str, defines the direction of the inference, options 'greater than' or 'less than'.  Default is 'greater than'.
        value: float,  defines the value about which to make the inference.  Default = 0.
        print_inference:  boolean, prints a readable string.  Default is True.
        names:  list of length 3, parameter names in order: a, b, b-a.  Default ['theta_a', 'theta_b', 'delta']
        Returns
        -------
        tuple
            - float, probability that b > (a + value) or b < (a + value).
            - str, string interpretation of that probabiliyu       
        '''
        dir_opts = ['greater than', 'less than']
        if direction not in dir_opts:
            raise ValueError("direction must be 'greater than' or 'less than'")
        if direction == 'greater than':
            p = len(self.d_draw[self.d_draw > 0]) / len(self.d_draw)
        else:
            p = len(self.d_draw[self.d_draw < 0]) / len(self.d_draw)
        i = self._probability_interpretation_guide(p)
        if names is None:
            names = [
                    'theta_a',
                    'theta_b',
                    'delta'
                    ]
        if len(names) > 3:
            raise ValueError('names must be a list of length 3')
        if print_inference is True:
            print(self._print_inference_probability(p, i, direction, value, names))
        return p, i

    def _bayes_factor_interpretation_guide(self, bf):
        # Interpretation guide for bayes factors using: 
        # Jeffreys guide (https://en.wikipedia.org/wiki/Bayes_factor#cite_note-9)
        if np.isinf(bf) or bf > np.power(10, 2):
            i = 'decisive'
        elif bf > np.power(10, 3/2) and bf <= np.power(10, 2):
            i = 'very strong'
        elif bf > 10 and bf <= np.power(10, 3/2):
            i = 'strong'
        elif bf > np.power(10, 1/2) and bf <= 10:
            i = 'substantial'
        elif bf >= 1 and bf <= np.power(10, 1/2):
            i = 'barely worth mentioning'
        elif bf < 1:
            i = 'negative'
        else:
            raise ValueError('bf did not satisfy range of criteria')
        return i

    def _print_inference_bayes_factor(self, bf, i, direction, value, names):
        s = 'The calculated bayes factor for the hypothesis that ' + names[1] + ' is ' + direction + ' ' + names[0]
        if value != 0:
            s = s + ' by more than ' + str(value)
        s = s + ' versus the hypothesis that ' + names[0] + ' is ' + direction + ' ' + names[0]
        if value != 0:
            s = s + ' by more than ' + str(value)
        s = s + ' is ' 
        if np.isinf(bf) is True:
            s = s + 'more than 100'
        else:
            s = s + ('%.5g' % bf)
        s = s + '. Therefore the strength of evidence for this hypothesis is ' + i + '.' 
        return s

    def _estimate_bayes_factor(self, p_h1, p_h2):
        # Estimates bayes Factor
        if p_h2 == 0:
            k = np.Infinity
        else:
            k = p_h1 / p_h2
        return k

    def infer_delta_bayes_factor(self, direction='greater than', value=0, print_inference=True, names=None):
        '''
        Provides a guide to making inferences on the posterior delta, based on the Bayes Factor by estimating
        P(D|H1) / P(D|H2) for the hypotheses H1: b>(a + value) vs H2: (a + value)>b (or vice versa).  
        Where D denotes the observed data.
        Parameters
        ----------
        direction: str, defines the direction of the inference, options 'greater than' or 'less than'.  Default is 'greater than'.
        value: float,  defines the value about which to make the inference.  Default = 0.
        print_inference:  boolean, prints a readable string.  Default is True.
        names:  list of length 3, parameter names in order: a, b, b-a.  Default ['theta_a', 'theta_b', 'delta']
        Returns
        -------
        tuple
            - float, bayes factor for P(D|H1) / P(D|H2) for the hypotheses H1: b>(a + value) vs H2: (a + value)>b (or vice versa).
            - str, string interpretation of that bayes factor    
        '''
        dir_opts = ['greater than', 'less than']
        if direction not in dir_opts:
            raise ValueError("direction must be 'greater than' or 'less than'")
        if direction == 'greater than':
            p_h1 = len(self.d_draw[self.d_draw > value]) / len(self.d_draw)       
            p_h2 = 1 - p_h1
            bf = self._estimate_bayes_factor(p_h1, p_h2)
        else:
            p_h1 = len(self.d_draw[self.d_draw < value]) / len(self.d_draw)       
            p_h2 = 1 - p_h1
            bf = self._estimate_bayes_factor(p_h1, p_h2)
        i = self._bayes_factor_interpretation_guide(bf)
        if names is None:
            names = [
                    'theta_a',
                    'theta_b',
                    'delta'
                    ]
        if len(names) > 3:
            raise ValueError('names must be a list of length 3')
        if print_inference is True:
            print(self._print_inference_bayes_factor(bf, i, direction, value, names))
        return bf, i

    def posterior_plot(self, 
                       method='hdi', 
                       delta_line=0,
                       col='#1f77b4', 
                       bounds=None,
                       names=None,
                       fig_size=None):
        '''
        Plots the density of the draws from the posterior distribution
        Parameters
        ----------
        method: str, defines method for interval estimate and central tendency.  Default = 'hdi'
            - 'hdi':  Uses HDI and maximum aposteriori
            - 'quantile': Uses credible intervals and median
        delta_line: float, position of the vertical line on the delta plot
        col:  str, colour of plots.  Default = '#1f77b4' (muted-blue)
        bounds:  float or list, defines the boundaries of the interval
            - if method = 'hdi': float, defines the interval of the HDI. Default = 0.95
            - if method = 'quantile': list, defines the credible interval.  Default = [0.025, 0.975]
        names: list of length 3, parameter names for the plot.  Default ['theta_a', 'theta_b', 'delta']
        fig_size:  tuple(width, height), dimensions of plot.  Default is None
        '''
        valid_methods = ['hdi', 'quantile']
        if method not in valid_methods:
            raise ValueError("method must be 'hdi' or 'quantile'")
        if method == 'hdi' and bounds is None:
            bounds = 0.95
        if method == 'quantile' and bounds is None:
            bounds = [0.025, 0.975]
        if method == 'hdi' and (bounds <= 0 or bounds >= 1):
            raise ValueError("if method is 'hdi' then bounds must be a float between 0 and 1")
        if method == 'quantiles' and len(quantiles) != 2:
            raise ValueError("quantiles must be a list of length 2")
        if names is None:
            names = [
                    'theta_a',
                    'theta_b',
                    'delta'
                    ]
        if len(names) > 3:
            raise ValueError('names must be a list of length 3')   
        if method == 'hdi':
            interval_name = 'hdi'
            centre_line_name = 'map'
        else: 
            interval_name = 'credible interval'
            centre_line_name = 'median'               
        fig = make_subplots(rows=1,
                            cols=3,
                            shared_xaxes=False,
                            shared_yaxes=False,
                            subplot_titles=tuple(names))
        draws = [
                 self.a_draw,
                 self.b_draw,
                 self.d_draw
                 ]
        for i in range(0,3):
            cl = _get_centre_lines(draws[i], method=method)
            intervals = _get_intervals(draws[i], method=method, bounds=bounds)
            fig.add_trace(_make_density_go(draws[i], name='posterior density', col=col), 1, i+1)
            fig.add_trace(_make_histogram_go(draws[i], name='posterior draws', col=col), 1, i+1)  
            fig.add_trace(_make_line_go(cl, name=centre_line_name, col=col), 1, i+1)   
            fig.add_trace(_make_area_go(intervals, name=interval_name, col=col), 1, i+1)       
        fig.update_layout(shapes=[_make_delta_line(self.d_draw, delta_line=delta_line)])
        fig.update_yaxes(title_text='density', row=1, col=1)
        name_set = set()
        fig.for_each_trace(lambda trace: 
            trace.update(showlegend=False)
                if (trace.name in name_set) else name_set.add(trace.name))
        if fig_size is not None:
            fig.update_layout(height=fig_size[1], width=fig_size[0])
        return fig





